### DISCUSSION TITLE
Turn OpenCart headless

### DISCUSSION THREAD

**Discussion by danielkerr** [opencart/opencart#13788] [Ideas]
I'm planning on turning opencart headless.

i think i can get the system to generate the main catalog pages into  headless html pages. 

I could add a folder called html and another called json then use htacess to redirect to those pages if match seo keyword path found.

so should generate:

html/category-1/product-1.html

and

html/category-1/sort-asc/page-1/product-1.html

and for json we use just generate files like json/country.json.

**Comment by felipoantonoff** [comment]
@danielkerr The idea is interesting, in my E-commerce it was 100% Vue.js + API.
We later switched to Twig, due to some problems with using just JS + API.

I consider Twig more powerful and I find it interesting to be able to use PHP, variables already filled in or API via JS, so the best one is used in each case.

A problem with JS + API is that it will expose all API calls, unlike Twig which may come with certain values already filled in by PHP, depending on the API it is important not to be publicly exposed.

Another point, there may be an excess of Requests, if a page needs to call about 5 resources, there will be 5 requests, so there is a delay in reaching the server, processing and returning, even though it can be worked in a more asynchronous way, a A good example of this occurs in Checkout, as it needs to be as realtime as possible, but in this case it makes sense to use more APIs with JS.

You also have to analyze the use of modules, you will lose a lot of modules, a lot of themes will not be compatible and other details.

Now I believe it is important for Opencart to have a good API, whether using HTML/JS or PHP + Twig.

In the example mentioned, it is perhaps more efficient to use a cache in Redis or another type of cache, I noticed that Opencart has a lot of the same SQL calls, on a page I believe there are more than 20 calls, most of which are resources such as currency, countries, settings etc., and all this data could come from a cache or assembled in such a way that a call already pulls all the data and can then be used from the global variables.

A good use of direct HTML I consider for Blog/CMS and Information Pages, in which it is possible to generate static pages, perhaps converting Markdown to HTML.

As for pages that require a lot of resources and modules, I believe the current PHP + Twig standard with HTML and JS being used as needed is more advantageous.

A cool use perhaps for HTML/JS + API would be for applications, perhaps a simplified version of Opencart for Mobile, but I believe it would require a lot of work and could be something to be focused on perhaps more by Modules or freelancer.

If your example is more for some locations, it might make sense for optimization, but I believe that the final HTML could be generated directly without needing json, similar to static CMS pages, already generating the final result, like a store that doesn't have a module on the product page you could maybe use.

In general, I consider Opencart to have excellent performance, we had fantastic results on a giant website with thousands of orders, which was at its peak consuming 10-16 CPU cores and thus slowing down the store, after using index in the database, Redis cache In general, the Redis session began to consume less than 1 CPU in the server's load average for the same demands, so Opencart can handle high demand very well with small optimizations.

**Comment by hayden-t** [comment]
what are the advantages ?

**Comment by danielkerr** [comment]
@FelipoAntonoff everything u have pointed out i agee with. so many issues with some of the frameworks liek even sea4rch engines will have issue indexing. 

i can make it html headless sites by caching all the main pages. certainly moving all the json to there own areas pre  rendend will speed up sites a lot.

**Comment by hayden-t** [comment]
Hi, im thinking about splitting the opencart site i dev into something 'headless' to satisfy the demands coming from the front end design and seo aspects of the business.

Im brainstorming ways to have a front end site (what ever they want eg wordpress) and backend site (opencart)

At one end of the spectrum I guess you could just use opencart api json and then let the front end site build everything else. Thats pretty much already done and possible but for adding a few more endpoints.

But im thinking about another option where opencart can serve up html content the way it currently does but without a html head, so the output can be inserted into the body of the front end site by itself, like opencart is acting as a addon/plugin 

Ive used an opencart plugin for joomla before, but it mostly sucked as the 2 systems werent independent enough.

One of the hurdles to deal with this approach is working out a method such that requests to the front end site are forwarded onto the backend site.

**Comment by danielkerr** [comment]
im planing to switch to headless also. im writing web components that will do a lot of the work.

**Comment by danielkerr** [comment]
maybe setup a generator page that will run and start generating templates and we use htacess to read the html page before hitting the php

**Reply by hayden-t** [reply]
are you suggesting to use htaccess to handle the routing to the generator ?

**Comment by hayden-t** [comment]
thanks for the update ! I had wondered what your recent thought were on it since last post but did ask specifically :)

**Comment by danielkerr** [comment]
also going to use nunjucks

**Comment by danielkerr** [comment]
i started some here https://github.com/opencart/opencart/tree/development

look under view. new folders are

view
 -- data // store JSON files
 -- html stored // pre generated html
 -- JavaScript 
 -- Stylesheet
 -- Template

**Reply by hayden-t** [reply]
i just see 
![image](https://github.com/user-attachments/assets/4331b2e9-546a-4b69-8776-acba1d2642ba)
just one data but no html folder

**Comment by hayden-t** [comment]
Is the idea that the 'generator' will generate html output the way the current site does but save it as files rather than send it as a response. And this generated html will be in elements/blocks , rather than one whole page ?

**Reply by mhcwebdesign** [reply]
> Is the idea that the 'generator' will generate html output the way the current site does but save it as files rather than send it as a response. And this generated html will be in elements/blocks , rather than one whole page ?

Won't make much sense if the generated HTML also involves dynamic variables? How would you handle that?

**Comment by danielkerr** [comment]
What about using htmx? Seems it saves on a lot of repeated code.

We can generate all pages as html or split in to segments. 

I want to remove the number of connections to the db. We can skip this by having JavaScript directly accessing is n files.

**Reply by hayden-t** [reply]
So the idea is to move some database content into json files for client side access ? Some db content is already cached for php to access without db

**Reply by danielkerr** [reply]
> So the idea is to move some database content into json files for client side access ? Some db content is already cached for php to access without db

Yes. We can do all products, categories, manufacture, Information, article and topic data

**Reply by punitkumar** [reply]
Any thought on Typescript or Rust or React ?

**Reply by felipoantonoff** [reply]
In this case, I believe that using Cache is even better than files. Maybe put Redis in front or leave it as it is today. Each person chooses the Cache driver. I only think files are bad when they involve a lot of use, because they can overload the disk with IO.

At a large client, we put a lot of pages in Cache in Redis to avoid multiple calls to MySQL, especially for configuration, currency, language, etc.

There could be a trigger that clears the cache in general or just its prefix when some content changes. For example, for the product cache, something like oc_product_... could be in front. Then, when a product is edited, everything with oc_product is cleared. If it is a setting, it could be oc_setting_... and so on. Basically, the table name is used as a prefix.

About HTMX https://htmx.org/ a colleague wrote a good article studying its use in his project together with Alpine, in this case he will create a Panel for his PaaS and Container system: https://ntorga.com/full-stack-go-app-with-htmx-and-alpinejs/

In place of alpline.js you can even use direct JS depending on what is done as well.

One detail has to be seen how the search works, but then I believe it could be a realtime Search API with the Bank or even if you already have the same word recently searched in the Cache you can use the cache. For searching I have used Full Text Search in an App/plugin, but for a client I needed to do a hybrid, I use LIKE + Full Text Search, because it first prefers to sort by what LIKE finds and then sort by the number of repeated words separately from Full Text Search.

**Comment by felipoantonoff** [comment]
@danielkerr Regarding database connections, I consider Twig to be better. In a project with Vue.js, we had to make many API calls, but of course there are ways to improve it, such as the suggested idea of ​​perhaps generating Json or something similar.

One detail is that depending on the API, we can put Redis or another Cache system in front, for example, Opencart settings, language, countries and other data that are loaded with each request.

As for a direct file generated, such as a Json, depending on the demand, there may be an IO bottleneck. In a large project in Opencart 2.3, we had a problem with the Cache in files. After switching to Redis, it went back to normal and in the database we used Index. When using Index, an E-commerce that depended on 16 CPU cores started to consume 1-2 CPUs or less and with high demand.

Perhaps Opencart could use Redis as the main Cache Driver and the API could be saved in it or another active cache driver. It is also possible to combine more than one API in a single call, but the options and complexity would need to be carefully analyzed.

Wouldn't it be better for Opencart to initially focus on having a good Rest or GraphQL API? With a good API, anyone can focus on creating an isolated Design. Then, Opencart could focus on a theme that handles these APIs or Json generators.

The biggest detail needs to be analyzed regarding SEO if it is a page like Home, Product, Category, Blog and others without Login and also the complexity of making modifications using OCMOD, for example. Today, with Twig, we have many resources and a single request that involves N database operations.

I would avoid using too much JS, because it is more complex to maintain and sometimes more expensive than a simple Twig with HTML + PHP + CSS and small JS. We have to see if it would be appropriate for Opencart to maintain 2 themes, one for Twig and another for the API, or if it will only be one. Perhaps headless makes more sense for the Login and internal areas, such as Checkout, Cart, My Account, or even to create an Opencart Mobile App for Android or Apple.

So, the main points to be analyzed:
- Performance and resource consumption
- Complexity to create and maintain
- Whether or not it will affect SEO
- Complexity for a Developer to create improvements that make visual changes, such as adding a Notify Me button on the Product page, for example
- Whether it will increase the dependencies too much
- Security of how it will handle the session and data that may be exposed in the API if the call is at the JS level

**Reply by danielkerr** [reply]
React is not fit for anything. Just Facebook promoting crap.

**Reply by felipoantonoff** [reply]
Yes, I also don't recommend using React, in fact I'm in favor of using less JS the better overall, using it more where it's really important like Cart, Checkout and other places where we unfortunately have to use JS.

**Comment by hayden-t** [comment]
Regarding my concept of 'headless/plugin' https://github.com/opencart/opencart/discussions/13788#discussioncomment-12166755 
Ive realised i only have to do this for the product and category pages, so they can load within another front end website body, everything else can run off sub.domain.com as they are not search engine accessible. This means 98% of the routing can be left as is. 
This make it much easier and save time, so OC just does the cart/account functionality natively, with product/category data passed to the front end website that host/manages everything else.
This way I can just run the technical backend and cart side of things, and the front end dev/s go do what ever they want to the front end site , picking what ever cms they want, and im not responsible or have to be involved.

**Reply by hayden-t** [reply]
oh, i can actually even just run OC in an iframe, so i get to keep(be presented in) the front end site header and footer, as once again iframe not indexed but doesnt matter as areas not accessible by crawlers

**Reply by felipoantonoff** [reply]
@hayden-t We did something similar in our previous project, it was like secure.dominio.com.br, in which a JS + API is used just for my account, checkout, shopping cart, etc.

I believe that the best use of JS is on dynamic pages like these that require a Login, while the other pages can be kept in the traditional way with Twig or, if you prefer, you can build your View with JS like Vue.js, React, etc.

Nowadays, I prefer Twig + JS, but all options are viable as long as they don't affect SEO or are used on pages that search engines should index.

The Admin can also be done with just JS + API, even because it requires a Login, but I believe that it may affect the application of improvements made in OCMOD to the View, unless OCMOD can also access JS files.

**Comment by eozalp** [comment]
İt will be great to separate frontend and backend. Making it more modular and more decoupled is the best way to keep it's relevance i think. Simple, modular, decoupled.

**Reply by felipoantonoff** [reply]
In theory, yes, it seems better, but in practice it depends a lot on how it is done, as it can have more dependencies and complexities.

Just like the Microservices trend, people have been talking badly about monolithic systems, but to this day the best systems in general are monolithic, especially because they meet more than 90% of use cases. Not everyone has a billion-dollar company that needs to meet millions of accesses at the same time, and then it can be worth dividing an application into small pieces.

In Opencart today, the view is already separate from the Back End, but it uses Twig, which is nothing more than simplified HTML + PHP and receives data from the controller.

Creating a 100% JS view that uses an Opencart API or even a Json has its pros and cons, and perhaps places that are more interesting to use, such as in the Checkout, cart, etc.

**Comment by fmarslan** [comment]
Hello,
I would like to share an idea to take OpenCart's headless structure further. In the current discussion about headless OpenCart, we can suggest turning OpenCart into a fully API-based system and returning dynamic templates as JSON to the client.
The main goal of this development is to send all the data to the UI in JSON format as a response, working like a dynamic template. The template data will be processed and returned in JSON format, making it easy to use with APIs. The system uses the current user authorization structure, so there is no need for extra authorization for API requests. Requests include a session ID and user token, allowing the system to behave as if it's a regular browser request.

**Key Features:**
**Authorization Integration:** The current user authorization system will be used for API requests. Users are already authorized when they log in, so no additional authorization steps are needed.
**Admin Panel Control:** In the admin panel, we can set which users are allowed to make API requests and add IP restrictions.
**Headless Setup:** The admin and UI panels can be used headlessly, with all actions happening through APIs.
This development will help turn OpenCart into a fully API-based system, making it more flexible. If this idea is accepted, we can submit the changes via a pull request (PR).

**Important!!**
We can manually remove data keys like header and footer in the API response, but all the data will be serialized. This can be seen as both a pro and a con.

Sample Result
![image](https://github.com/user-attachments/assets/ef74d61f-0ade-4ac9-934d-f5711b2e73cd)


Looking forward to your feedback!
Thank you!

**Reply by mhcwebdesign** [reply]
> Hello, I would like to share an idea to take OpenCart's headless structure further. In the current discussion about headless OpenCart, we can suggest turning OpenCart into a fully API-based system and returning dynamic templates as JSON to the client. 
> 
> Looking forward to your feedback! Thank you!

And how do you suggest the client-side to be implemented with these JSON-data from the API calls?

Aren't we just re-inventing the wheel here?

**Reply by fmarslan** [reply]
1. How do you suggest the client-side to be implemented with these JSON-data from the API calls?
The client-side can use Twig(built-in feature) for rendering or modern frameworks like Vue or React for dynamic UIs. This gives flexibility to choose the best tool for the job.

2. Aren't we just re-inventing the wheel here?
Not exactly. The goal is to make OpenCart more flexible and API-driven, allowing easy integration with different front-end technologies and third-party services.

Many pages already have data in JSON format, but as you mentioned, this data is often customized for the UI (like edit variant action links or sometimes splitting the data). The real benefit here is being able to directly access important data like product details or order details, which can already be accessed right now. In the future (maybe in the next version), if we pass this data to Twig within a wrapper object, we can easily remove unnecessary parts centrally and get cleaner data. This way, we won’t need to develop or manage extra APIs. The UI can use Twig rendering if needed, or work with modern JavaScript frameworks like Vue or React, or even integrate with third-party services.

Right now, the biggest problem is the form data sent via POST from the UI. The naming and hierarchy of this data can sometimes cause issues. But in the future, if we create a standard for this, any validation or updates made for the UI will automatically apply to the APIs as well.

**Reply by felipoantonoff** [reply]
Interesting, the idea from what I understand is to avoid having redundancy when dealing with data.

Using only one API that can be used directly by Twig for its data or sent as Json to be used by JS view or to a service that only wants to use the API.

Opencart does a bit of this for certain Administrator features, but the API part and the part that passes the data to the view are well separated.

I think it would be interesting if it were to simplify and have only one place for assembling the data and of course also for performing the appropriate controller actions, which would basically be an API.

What Opencart needs is a RESTful API basically, when it involves assembling the view or using it as a traditional API, a json can be used and if it is to pass to Twig an array for example, avoiding redundancy.

I think that before the visual focus it is more important to have a good API used by the catalog and administrator for all the main actions and data, also implemented with a rate limit to avoid possible abuse.

Once you have a good API in Opencart, which can be used for general purposes and for external services connected to the API, it may be worth thinking about the visual aspect and possible reformulation or, by default, keeping Twig and improving the current theme. You can give an example of use with more JS in the Checkout, Cart and My Account sections, which work well for screens that depend more on JS.

In reality, the API can always return json, but when it is called to be used in Twig, there should only be an auxiliary method, either in the next call to do json_decode or in the call itself, passing a variable for this.

It would also be good to have the cache up front and for this cache to be cleared when data changes are involved, such as new, updated or removed data.

Internally, to use the API, it would be enough to call perhaps the direct method or even a simple Curl for internal calls.

**Comment by mhcwebdesign** [comment]
> The client-side can use Twig(built-in feature) for rendering or modern frameworks like Vue or React for dynamic UIs. This gives flexibility to choose the best tool for the job.

Twig is a rendering engine which runs on the server, not on the client's end.

**Reply by fmarslan** [reply]
Twig is a template engine that is already built into the server and allows you to customize the user interface (UI) with different templates.

Imagine each page has a parameter like api=true at the end, or you're redirecting to a file like api.php instead of index.php. When a request like this is made, the Twig engine is turned off, some data keys are cleared, and the user gets a response in JSON format. Similarly, incoming POST requests are forwarded directly to the relevant controller, acting like a proxy.

For example, in the current OpenCart setup, data is directly passed to the output using an event like this:

```php
public function before(&$route, &$data, &$code, &$output):void{
    // If the request route is "extension/opencart/api/select"
    if(isset($this->request->get['route']) && 
        $this->request->get['route'] == "extension/opencart/api/select") {
        // The data is encoded in JSON format and assigned to the output
        $output = json_encode($data);
    }
}
```

**Comment by fixthemad** [comment]
There is an OC 4.0 fork that aims to be a headless option for OpenCart here: https://github.com/captainerd/VentoCart. The frontend is React-based and doesn't seem to be complete yet, but it may be a promising project.
